### PR TITLE
Isolated device factory

### LINKED ISSUE #483: Make `device_factory` decorator to ease making and connecting ophyd-async Devices
https://github.com/DiamondLightSource/dodal/issues/415#issuecomment-2079651830 summarizes a discussion on Device connection.
https://github.com/bluesky/ophyd-async/issues/265 will implement the ophyd-async supporting features of idempotent connect. The strategy we would like for dodal is:
- Each beamline module decorates its device factories with a `device_factory`
- This factory makes, names, caches and optionally connects a device
- In script use can pass `connect=True` to the device to connect in the foreground, requires a RunEngine to have been made
- In blueapi use can find all the eager devices and connect them at startup
- Can also connect them in a plan with the `ensure_connected` plan stub

This ticket is for:

## New ADR
Describing how Devices should be connected in dodal, from scripts and under blueapi

## `dodal.beamlines.beamline_utils`
```python
D = TypeVar(bound=Device)
class DeviceFactory(Protocol, Generic[D]):
    def __call__(self, connect: bool = False, timeout: float = DEFAULT_TIMEOUT) -> D:
        ...

_factory_made_devices: Dict[DeviceFactory, Device] = {}
_device_is_lazy: Dict[DeviceFactory, bool] = {}

def device_factory(lazy=False, set_name=True):
    def wrapper(f) -> DeviceFactory[D]:
        def factory(connect=False, timeout: float = DEFAULT_TIMEOUT):
            device = _factory_made_devices.get(f)
            if not device:
                device = f()
                if set_name:
                    device.set_name(f.__name__)
                _factory_made_devices[f] = device
            if connect: 
                call_in_bluesky_event_loop(device.connect(timeout=timeout))
            return device
        _device_is_lazy[factory] = lazy
        factory.__name__ = f.__name__
        return factory
    return wrapper

def get_device_factories() -> Dict[DeviceFactory, bool]:
    return _device_is_lazy.copy()
```
Need to add type hints to this so it knows the type of device it is making from the function it is decorating and check with @callumforrester that `get_device_factories` is what blueapi actually needs

## Acceptance Criteria
* ADR written so we know why this was done
* `device_factory` available to use in beamline modules
* `i22.py` updated to use it

### PR DESCRIPTION
**Pull request by DiamondJoseph**
Fixes #483 

This is a re-implementation of #597 isolated to just the changes necessary to implement the `@device_factory()` decorator and have functions decorated with that decorator picked up by `make_all_devices`: respecting any configuration passed into the decorator.

NB: this is also required for [blueapi #647](https://github.com/DiamondLightSource/blueapi/issues/647)

### Instructions to reviewer on how to test:
1. System test the i22 beamline to ensure that devices are instantiated, connected and named appropriately.

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://diamondlightsource.github.io/dodal/main/reference/device-standards.html)
- [ ] If changing the API for a pre-existing device, ensure that any beamlines using this device have updated their Bluesky plans accordingly
- [ ] Have the connection tests for the relevant beamline(s) been run via `dodal connect ${BEAMLINE}`

### CODE REVIEW

**Comment by callumforrester** [comment]
I don't necessarily expect parallel connect to be implemented here but do we have tickets for achieving it?

**Comment by DiamondJoseph** [comment]
> I don't necessarily expect parallel connect to be implemented here but do we have tickets for achieving it?

https://github.com/DiamondLightSource/blueapi/issues/440

**Review by callumforrester — APPROVED** [review]

**Review by stan-dot — APPROVED** [review]
nice functools use

File: src/dodal/utils.py
Comment (DominicOram) [inline src/dodal/utils.py:153]: Could: This is basically the interface for when I do `i22.dcm(...)` from a script. As such it would be better to make the arg names a bit cleaner e.g. `connect_immediately`? and `connection_timeout`

Comment (DominicOram) [reply src/dodal/utils.py:153]: Also docstrings but happy for that to be covered by https://github.com/DiamondLightSource/dodal/issues/780

File: src/dodal/utils.py
Comment (DominicOram) [inline src/dodal/utils.py:101]: Should: Can we call this something like `skip_type` to avoid confusion?

File: src/dodal/utils.py
Comment (DominicOram) [inline src/dodal/utils.py:275]: Could: Is `fake_with_ophyd_sim` purely in there for backwards compatibility downstream? I think from an MX POV we can break compatibility here and it will be fine, we will need to do a bit of migration work anyway.

File: src/dodal/utils.py
Comment (DominicOram) [inline src/dodal/utils.py:312]: Should: Can we have a test around this please?

Comment (DominicOram) [reply src/dodal/utils.py:312]: Could: It's not 100% clear what the difference is between this and `collect_factories` and when one should use this or the other. It would be good to have some more in the docstrings or change the function names to make it clearer

File: src/dodal/utils.py
Comment (DominicOram) [inline src/dodal/utils.py:351]: Should: Can we have a test here please?

**Review by DominicOram — CHANGES_REQUESTED** [review]
Great, thank you!

Some comments in code, mostly minor. I think a few more tests would also be good. e.g. we can test (assuming these are all desired behavior?):
* A device gets renamed if called again with a different name 
* If we call with `connect=False, mock=True` we can then call again with `mock=False`

Additionally I think there's a potential pitfall in:
```python
dcm(mock=True)
.... # Lots of code
dcm() # This now fails because it tries to turn the cached mock one into a real one
```
I'm not 100% sure if this is expected or if you would rather give the cached mocked one, maybe with a warning. We should at least document it but can do so in #780.

File: src/dodal/utils.py
Comment (DiamondJoseph) [reply src/dodal/utils.py:275]: Yes that's there for backwards compat- is dodal still wanting a "make all devices" method (just for `dodal connect`?) or is that API going to move into blueapi? I was going to make a PR after this got merged to remove the kwargs and pass explicitly the args that connect wants (after moving more beamlines over).

File: src/dodal/utils.py
Comment (callumforrester) [reply src/dodal/utils.py:153]: As an aside: I've been in enough debates about verbosity vs. "why say many word when few do trick" now to wonder if we need a style guide. Java is very pro-verobosity and has that culture around it so we never needed one with GDA, the Python world is a bit more fragmented.

File: src/dodal/utils.py
Comment (callumforrester) [reply src/dodal/utils.py:101]: Also maybe should be in PascalCase, similar to most other types and type aliases.

File: src/dodal/utils.py
Comment (callumforrester) [reply src/dodal/utils.py:275]: I'm also happy to break compatibility, however this change is required to fix some ongoing blueapi build issues so maybe we shouldn't block it/do that separately?

https://github.com/DiamondLightSource/blueapi/issues/665
https://github.com/DiamondLightSource/blueapi/issues/647

Comment (DominicOram) [reply src/dodal/utils.py:275]: Yes, we still want `make_all_devices` I think. If it's going to be addressed in another issue that's fine.

File: src/dodal/utils.py
Comment (callumforrester) [reply src/dodal/utils.py:312]: Hmm, they are very similar, maybe we should just mark `collect...` as deprecated and have done with it?

File: src/dodal/utils.py
Comment (DominicOram) [reply src/dodal/utils.py:153]: Ooof, it'll be controversial and I think hard to codify into a style guide. I feel like the answer is always "As verbose as is required to make it clear given the context" which is so subjective.  But I would be happy to weigh in on a discussion

File: src/dodal/utils.py
Comment (DominicOram) [reply src/dodal/utils.py:312]: Marking as depreciated works for me, with an explaining comment/link to issue

File: src/dodal/utils.py
Comment (DominicOram) [reply src/dodal/utils.py:101]: > Also maybe should be in PascalCase, similar to most other types and type aliases.

I also agree with this but thought it maybe too opinionated to add. I would rather PascalCase though.

File: src/dodal/utils.py
Comment (callumforrester) [reply src/dodal/utils.py:153]: It is quite subjective, but it's actually quite useful. What you said is a nice middle ground between "you should minimise the amount of lines/characters you write" (C) or "You should write function signatures as full sentences" (Java)

File: src/dodal/utils.py
Comment (DiamondJoseph) [reply src/dodal/utils.py]: https://github.com/DiamondLightSource/dodal/issues/844

File: src/dodal/utils.py
Comment (DiamondJoseph) [reply src/dodal/utils.py]: I just nixxed this implementation and reworded the ADR: the only difference was that it returned a dict[str, DeviceInitilizationController] rather than a dict[str, Callable[[], AnyDevice]: as beamlines move over this can be converted to all return DIC, and when we call it from blueapi we can check if it's a DIC and handle appropriately

File: src/dodal/utils.py
Comment (DiamondJoseph) [reply src/dodal/utils.py]: :sunglasses: :+1:

File: src/dodal/utils.py
Comment (DiamondJoseph) [reply src/dodal/utils.py]: :sunglasses: :+1:

File: src/dodal/utils.py
Comment (DiamondJoseph) [reply src/dodal/utils.py:369]: :sunglasses: :+1:

**Review by DominicOram — APPROVED** [review]
Thanks. I'm still a little concerned about the not 100% intuitive behavior of calling the factory twice with different permutations of connect/mock but I thin we can document/fix that later.

**Comment by DiamondJoseph** [comment]
> Assuming these are all desired behavior?

@coretl for comments?

>     * A device gets renamed if called again with a different name

[A device can be renamed after having a name](https://github.com/bluesky/ophyd-async/blob/3e74761b76f337bf0e36dcdbdd7fdc0dd814105b/src/ophyd_async/core/_device.py#L55), so I think it should be allowed. Currently the device_factory name is only used if the device is currently unnamed.

>     * If we call with `connect=False, mock=True` we can then call again with `mock=False`

Only when connect is called is the device being mock or not considered. Calling with connect=False, mock=True then later connecting with mock=False would connect to the real signal backend.

> Additionally I think there's a potential pitfall in:
> ```python
> dcm(mock=True)
> .... # Lots of code
> dcm() # This now fails because it tries to turn the cached mock one into a real one
> ```

[Ophyd-async throws an exception if you try and connect a 2nd time with a different value of `mock`.
](https://github.com/bluesky/ophyd-async/blob/3e74761b76f337bf0e36dcdbdd7fdc0dd814105b/src/ophyd_async/core/_device.py#L96). I will amend to cache the value of `mock` after `connect` has been called at least one, and defer to it rather than the `mock` value set up with the decorator, so that we can still try and explicitly connect to [non-]mock, but will throw the exception only when we do so.

**Comment by DominicOram** [comment]
> I will amend to cache the value of mock after connect has been called at least one, and defer to it rather than the mock value set up with the decorator, so that we can still try and explicitly connect to [non-]mock, but will throw the exception only when we do so.

I'm not even sure that is the best answer either. You then end up with the above code connecting you to a `mock` when you actually want a real device and potentially you fail to realise. I'm not sure there's a good solution here. I think the best thing to do is just document that calling the factory multiple times can lead to odd behavior and is not recommended.

**Comment by callumforrester** [comment]
I agree there is no good solution to be honest, caching singletons is hard to do without some hidden magic. Maybe add more logging (but in a separate change).

**Pull request merged**
